File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];
    
    const n = maps.length;
    const m = maps[0].length;
    
    const visited = Array(n).fill().map(() => Array(m).fill(false));
    visited[0][0] = true;
    
    const que = [];
    que.push([0, 0, 1]);
    
    while(que.length > 0){
        const [x, y, w] = que.shift();

        if(x === n - 1 && y === m - 1){
            return w;
        }
        for(let d = 0; d < 4; d++){
            let nx = x + dx[d];
            let ny = y + dy[d];
            if(nx >= 0 && nx < n && ny >= 0 && ny < m){
                if(maps[nx][ny] === 1 && visited[nx][ny] === false){
                    visited[nx][ny] = true;
                    que.push([nx, ny, w + 1]);
                }
            }
        }
    }
    
    
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int[][] vMap;
    static int n, m;
    
    public int solution(int[][] maps) {
        n = maps.length;
        m = maps[0].length;
        Queue<int[]> que = new LinkedList<>();
        vMap = new int[n][m];
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                vMap[i][j] = Integer.MAX_VALUE;
            }
        }
        
        que.offer(new int[]{0, 0, 1});
        
        while(que.isEmpty() == false){
            int[] temp = que.poll();
            int x = temp[0];
            int y = temp[1];
            int value = temp[2];
            
            for(int d = 0; d < 4; d++){
                int nx = x + dx[d];
                int ny = y + dy[d];
                if(isIn(nx, ny) == false) continue;
                if(maps[nx][ny] == 1 && vMap[nx][ny] > value + 1){
                    vMap[nx][ny] = value + 1;
                    que.offer(new int[] {nx, ny, value + 1});
                }
            }
        }
                
        return vMap[n - 1][m - 1] == Integer.MAX_VALUE ? -1 : vMap[n - 1][m - 1];
    }
    
    public boolean isIn(int x, int y){
        return x >= 0 && x < n && y >= 0 && y < m;
    }
}